### PR TITLE
Allow to disable realtime=true per index

### PR DESCRIPTION
**Pull request by s1monw**
Today we force refresh on get calls if the document has changes that are
not yet refreshed. This might have implications and uses might want to disable
this to ensure cluster stability. We also can, if we don't need to serve realtime
requests at all optimize indexing to not populate the version map in such a case and
use more memory for the index writer buffers to speed up indexing.

Relates to #19813

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -331,6 +333,9 @@ public GetResult get(Get get, Function<String, Searcher> searcherFactory) throws
         try (ReleasableLock lock = readLock.acquire()) {
             ensureOpen();
             if (get.realtime()) {
+                if (isRealtimeGetEnabled == false) {
+                    throw new IllegalArgumentException("get with realtime=true is disabled on this index");
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: since realtime=true is the default, maybe the message should rather say something like "realtime gets are disabled. It is still possible to perform non-realtime gets by passing realtime=false"?

File: docs/reference/docs/get.asciidoc [hunk 2]
```diff
@@ -56,6 +56,10 @@ When getting a document, one can specify `fields` to fetch from it. They
 will, when possible, be fetched as stored fields (fields mapped as
 <<mapping-store,stored>> in the mapping).
 
+The realtime get can be disabled for an entire index by setting `index.enable_realtime_get` to
+`false`. This setting can not be changed at runtime. If realtime get is disabled
+all calls to the get API with the `realtime` flag set to `true` with fail.
```
Comment (jpountz) [inline docs/reference/docs/get.asciidoc]: Maybe rephrase to something like "If realtime get is disabled, only GET calls that pass `false` to the `realtime` flag will succeed." since realtime GET is the default

**Comment by jpountz** [comment]
I understand how the GET case works. However it's not clear to me what happens if you index a document with an auto-generated id and then immediately update/delete it since it will not be in the version map. Can you explain how those cases work?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Should we just default to false if realtime gets are disabled? I know we're removing all the crazy conditional defaults but this one doesn't seem so crazy because the user configures the index.

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Since the person who configures the cluster might not be the same as the one who sends GET requests, I think I like better that we make things explicit and require users to pass realtime=false.

File: docs/reference/docs/get.asciidoc [hunk 3]
```diff
@@ -56,6 +56,10 @@ When getting a document, one can specify `fields` to fetch from it. They
 will, when possible, be fetched as stored fields (fields mapped as
 <<mapping-store,stored>> in the mapping).
 
+The realtime get can be disabled for an entire index by setting `index.enable_realtime_get` to
+`false`. This setting can not be changed at runtime. If realtime get is disabled
```
Comment (nik9000) [inline docs/reference/docs/get.asciidoc]: s/This setting can not be changed at runtime./This setting may only be set at index creation./?

**Comment by s1monw** [comment]
> I understand how the GET case works. However it's not clear to me what happens if you index a document with an auto-generated id and then immediately update/delete it since it will not be in the version map. Can you explain how those cases work?

yeah so update doesn't work since RT get is disabled, if you trick ES into using the same ID for an update you have 2 documents with the same ID I guess. that is something that I was thinking about for a while and I am convinced now that I think this can't be done with this so I am rolling back the optimization and only stick with RT get disable

**Comment by jpountz** [comment]
Since disabling realtime GET on its own is not super useful, maybe this means we need a setting that disables realtime GET and only allows index operations with auto-generated ids at the same time (rejecting delete operations and index operations with explicit ids)?

**Comment by jpountz** [comment]
(I realize this might be much trickier than it sounds due to replicas or recoveries.)

**Comment by s1monw** [comment]
closing for now
